src/pal/pal_apple.h

```diff
@@ -1,92 +1,192 @@
+      puts("");
+      fflush(stdout);
+    }
 
     /**
-     * Note: The root's implementation works fine on Intel
-     * however mprotect/PROT_NONE fails on ARM
-     * especially since the 11.2 release (seems known issue
-     * spotted in various projects; might be a temporary fix).
+     * Report a fatal error and exit.
      */
-    template<bool page_aligned = false>
-    static void zero(void* p, size_t size) noexcept
+    [[noreturn]] static void error(const char* const str) noexcept
     {
-      if constexpr (Aal::aal_name != ARM)
-        PALBSD::zero(p, size);
-      else
-        ::bzero(p, size);
+      puts(str);
+      print_stack_trace();
+      abort();
     }
 
-#  if defined(PLATFORM_IS_ARM)
     /**
-     * Overriding here to mark the page as reusable
-     * rolling it as much as necessary.
-     * As above, the x86 h/w worked alright without this change
-     * however now large allocations work better and more reliably
-     * with on ARM, not to mention better RSS number accuracy
-     * for tools based on task_info API.
+     * Notify platform that we will not be using these pages.
      */
     static void notify_not_using(void* p, size_t size) noexcept
     {
-      SNMALLOC_ASSERT(is_aligned_block<PALBSD::page_size>(p, size));
-#    ifdef USE_POSIX_COMMIT_CHECKS
-      memset(p, 0x5a, size);
-#    endif
-      while (madvise(p, size, MADV_FREE_REUSABLE) == -1 && errno == EAGAIN)
-        ;
+      SNMALLOC_ASSERT(is_aligned_block<page_size>(p, size));
+
+      mach_vm_behavior_set(
+        mach_task_self(),
+        reinterpret_cast<mach_vm_address_t>(p),
+        mach_vm_size_t(size),
+        VM_BEHAVIOR_REUSABLE);
     }
 
     /**
-     * same remark as above but we need to mark the page as REUSE
-     * first
+     * Notify platform that we will be using these pages.
      */
     template<ZeroMem zero_mem>
     static void notify_using(void* p, size_t size) noexcept
     {
       SNMALLOC_ASSERT(
-        is_aligned_block<PALBSD::page_size>(p, size) || (zero_mem == NoZero));
-      while (madvise(p, size, MADV_FREE_REUSE) == -1 && errno == EAGAIN)
-        ;
+        is_aligned_block<page_size>(p, size) || (zero_mem == NoZero));
+
+      mach_vm_behavior_set(
+        mach_task_self(),
+        reinterpret_cast<mach_vm_address_t>(p),
+        mach_vm_size_t(size),
+        VM_BEHAVIOR_REUSE);
 
       if constexpr (zero_mem == YesZero)
+      {
         zero<true>(p, size);
+      }
+    }
+
+    /**
+     * OS specific function for zeroing memory.
+     */
+    template<bool page_aligned = false>
+    static void zero(void* p, size_t size) noexcept
+    {
+      if (page_aligned || is_aligned_block<page_size>(p, size))
+      {
+        SNMALLOC_ASSERT(is_aligned_block<page_size>(p, size));
+
+        // mask has least-significant bits set
+        mach_vm_offset_t mask = page_size - 1;
+
+        int flags =
+          VM_FLAGS_FIXED | VM_FLAGS_OVERWRITE | VM_MAKE_TAG(PALAnonID);
```

Comment:
would putting back VM_MAKE_TAG into a default flag work for you ?

Reply:
I'd rather leave it as an expression. If in the future Apple refactors VM_MAKE_TAG into a function or any other non constexpr it should still compile.